rep-2017.rst

```diff
@@ -0,0 +1,359 @@
+
+* Thread name
+* Core affinity
+* Scheduling policy
+* Priority
+
+When using the parameter passing feature, users should create an array containing sets of these items as thread attribute parameters for the thread pool used by ROS 2 executors.  The text format that users should comply with for describing parameters is YAML.  Users can choose to pass the parameters either as a string or from a file.
+
+Command line parameter
+''''''''''''''''''''''
+To use the command line parameter method, the users can utilize the following options:
+
+.. code-block:: bash
+
+    --thread-attrs-value=[Thread attribute parameters in YAML format]:
+      This option allows users to directly pass the parameters in YAML format as a command line argument.
+    --thread-attrs-file=[Path for the file including the thread attributes parameters in YAML format]:
+      This option allows users to specify a file path that contains the thread attribute parameters in YAML format.
+
+If multiple options are provided simultaneously, the first option encountered shall take precedence.
+
+Environment variables
+'''''''''''''''''''''
+To use the environment variable method, the user can utilize the following environment variables:
+
+.. code-block:: bash
+
+   ROS_THREAD_ATTRS_VALUE=[Thread attribute parameters in YAML format]:
+     Using this environment variable, users can pass the thread attribute parameters directly in YAML format.
+   ROS_THREAD_ATTRS_FILE=[Path for the file including the thread attributes parameters in YAML format]:
+     Using this environment variable, users can specify a file path that contains the thread attribute parameters in YAML format.
+
+If both are provided simultaneously, the ROS_THREAD_ATTRS_VALUE shall take precedence.
+
+Format of parameters in YAML
+''''''''''''''''''''''''''''
+
+Users should describe the thread attribute parameters in the YAML format having keys below.
+
+.. code-block:: YAML
+
+  priority:          Integer value expressing the priority of the thread
+  name:              String value representing the name of the thread
+  core_affinity:     Integer value used to decide the core on which the thread should run
+  scheduling_policy: String value specifying the desired scheduling policy for the thread
+
+The treatment of these values is environment-specific.  Specifically, for the `schduling_policy,` the following options are imported from the Linux environment and POSIX specification: [#REF-1]_ , [#REF-2]_
+
+.. code-block:: TEXT
+
+  FIFO
+  RR
+  SPORADIC
+  OTHER
+  IDLE
+  BATCH
+  DEADLINE
+
+..
+    Please note that the list above is subject to modification during the review process for this REP.
+    For instance, it may be necessary to include an option specifically designed for extension purposes in a non-POSIX environment.
+
+Using the provided keys, users should create an array of the thread attribute parameters like the one below.
+
+.. code-block:: YAML
+
+    - priority: 20
+      name: thread-1
```

Comment:
How user can know this thread does exist in the process space that is going to run? (e.g 3rd party developed node.)

and if it does not exist, this command line fails to start the application since it cannot set the requested thread attributes? or print the warning and fall back to normal thread attributes set? just curious how to behave this command line interface.

Reply:
First, in this REP's proposal, rcl does not generate any threads. rcl just receives the thread attribute parameters and stores in it to be retrieved from the language binding layers.

rcl only checks whether the thread attribute parameters passed are correct or not in terms of YAML grammar. It is the upper language binding layer to check if the parameters are acceptable for the environment or not.

For the existing applications you are concerned, there is no effect on them because the extension of this REP does not affect spontaneously. 
But, for those applications, my team will provide some extensions to enable them to use this infrastructure. In the proposal, the rclcpp's single/multithread executors define the reserved names for thread parameter names as follows:
- RCLCPP_EXECUTOR_MULTI_THREAD
- RCLCPP_EXECUTOR_SINGLE_THREAD

And, if they find the thread attribute having those names as string parameters, the executors use the parameters for the thread attributes used by the thread of each executor.
With this extension, the existing applications that do not use environment-specific thread control APIs will be able to use the newly added infrastructure.

Reply:
FYI

>But, for those applications, my team will provide some extensions to >enable them to use this infrastructure. In the proposal, the rclcpp's >single/multithread executors define the reserved names for thread >parameter names as follows:
>   - RCLCPP_EXECUTOR_MULTI_THREAD**ED**
>   - RCLCPP_EXECUTOR_SINGLE_THREAD**ED**

The extension above has been included in the commit below.

https://github.com/esol-community/rclcpp/commit/35af6fa9ff56cbd14b49650839b784e6d7db2cd3

And you can get started with the document below, using the benchmark applications (autoware_default_singlethreaded/autoware_default_multithreaded).

https://github.com/esol-community/reference-system/blob/rep2017_demo/misc/getting_started_thread_attribute_configuration_extention_rep2017.md